File: backend/app/schemas/resource.py
"""
资源相关的Pydantic schemas
"""
from typing import List, Optional
from pydantic import Field, field_validator

from app.schemas.base import BaseSchema, AuditSchema, PaginationParams
from app.schemas.tag import TagListResponse


class ResourceBase(BaseSchema):
    """
    资源基础模型
    """
    name: str = Field(..., min_length=1, max_length=255, description="资源名称")
    description: Optional[str] = Field(None, description="资源描述")
    tag_ids: List[str] = Field(default_factory=list, description="标签ID列表")


class ResourceCreate(ResourceBase):
    """
    创建资源模型
    """
    pass


class ResourceUpdate(BaseSchema):
    """
    更新资源模型
    """
    name: Optional[str] = Field(None, min_length=1, max_length=255, description="资源名称")
    description: Optional[str] = Field(None, description="资源描述")
    tag_ids: Optional[List[str]] = Field(None, description="标签ID列表")


class ResourceResponse(ResourceBase, AuditSchema):
    """
    资源响应模型
    """
    id: str = Field(..., description="资源ID")
    file_name: str = Field(..., description="原始文件名")
    file_size: int = Field(..., description="文件大小（字节）")
    file_type: str = Field(..., description="文件类型")
    user_id: str = Field(..., description="创建者ID")
    tags: List[TagListResponse] = Field(default_factory=list, description="资源标签")
    file_url: Optional[str] = Field(None, description="文件下载URL")


class ResourceListResponse(BaseSchema):
    """
    资源列表响应模型
    """
    id: str = Field(..., description="资源ID")
    name: str = Field(..., description="资源名称")
    file_name: str = Field(..., description="原始文件名")
    file_size: int = Field(..., description="文件大小（字节）")
    file_type: str = Field(..., description="文件类型")
    user_id: str = Field(..., description="创建者ID")
    tags: List[TagListResponse] = Field(default_factory=list, description="资源标签")
    created_at: str = Field(..., description="创建时间")


class ResourceSearchParams(PaginationParams):
    """
    资源搜索参数模型
    """
    keyword: Optional[str] = Field(None, description="搜索关键词")
    tag_ids: Optional[List[str]] = Field(None, description="标签ID列表")
    file_type: Optional[str] = Field(None, description="文件类型")
    user_id: Optional[str] = Field(None, description="创建者ID")

File: frontend/src/types/resource.ts
/**
 * 资源标签类型定义
 */
export interface ResourceTag {
  id: string;
  name: string;
  description?: string;
  color?: string;
}

/**
 * 资源文件类型
 */
export type ResourceFileType = 'image' | 'pdf' | 'doc' | 'video' | 'other';

/**
 * 资源类型定义 - 与后端 ResourceResponse 对齐
 */
export interface Resource {
  id: string;
  name: string;
  description?: string;
  file_name: string;
  file_size: number;
  file_type: string;
  user_id: string;
  tags: ResourceTag[];
  file_url?: string;
  created_at: string;
  updated_at: string;
}

/**
 * 资源列表项（简化版，用于列表展示）- 与后端 ResourceListResponse 对齐
 */
export interface ResourceListItem {
  id: string;
  name: string;
  file_name: string;
  file_size: number;
  file_type: string;
  user_id: string;
  tags: ResourceTag[];
  created_at: string;
}

/**
 * 资源创建请求数据 - 与后端 ResourceCreate 对齐
 */
export interface ResourceCreateData {
  name: string;
  description?: string;
  tag_ids?: string[];
  file: File;
}

/**
 * 资源更新请求数据 - 与后端 ResourceUpdate 对齐
 */
export interface ResourceUpdateData {
  name?: string;
  description?: string;
  tag_ids?: string[];
}

/**
 * 资源查询参数 - 与后端 ResourceSearchParams 对齐
 */
export interface ResourceQueryParams {
  page?: number;
  page_size?: number;
  keyword?: string;
  tag_ids?: string[];
  file_type?: string;
}

/**
 * 资源列表响应 - 与后端 ListResponse 对齐
 */
export interface ResourceListResponse {
  data: ResourceListItem[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

/**
 * 资源筛选状态
 */
export interface ResourceFilterState {
  searchText: string;
  selectedTags: string[];
  selectedType: ResourceFileType | null;
  currentPage: number;
  pageSize: number;
}
